app/controllers/proposals_controller.rb

```diff
@@ -23,6 +23,35 @@ def archive
     @proposals = self.chronological_proposals.closed
   end
 
+  def cancel_form
+    @proposal = self.proposal.decorate
+    unless policy(@proposal).can_cancel?
+     redirect_to proposals_path, alert: "You are not allowed to perform that action"
+    end
+
+    if @proposal.cancelled?
+     redirect_to proposal_path, id:@proposal.id, alert: "This request that has already been cancelled."
+    end
```

Comment:
You shouldn't need an explicit check on `can_cancel?` -- that should be automatically checked when `authorize`ing. Further, `can_cancel?` should probably both check that the user is allowed to perform the action and that the proposal isn't cancelled already. See `ProposalPolicy#can_edit!` for an example.

You'll still need the redirect logic, however; see examples w/ `auth_errors`. You can check `exception.query` (which would be a symbol).